octoprint_firmwareupdate/__init__.py

```diff
@@ -1,85 +1,271 @@
-    		    if 'error:' in line:
-    			    error_list.append(line)
-    	    compileError = '<pre>' + ''.join(error_list) + '</pre>'
-    	    self._logger.info("Update failed. Compiling error.")
-    	    self.isUpdating = False
-    	    self._plugin_manager.send_plugin_message(self._identifier, dict(isupdating=self.isUpdating, status="failed", reason=compileError))
-    	    return False
-    	elif 'Make failed' in update_result:
-    	    self._logger.info("Update failed. Compiling error.")
-    	    self.isUpdating = False
-    	    self._plugin_manager.send_plugin_message(self._identifier, dict(isupdating=self.isUpdating, status="failed", reason="Build failed."))
-    	    return False
-    	else:
-    	    self._plugin_manager.send_plugin_message(self._identifier, dict(isupdating=self.isUpdating, status="continue"))
-    	    return True
+        while True:
+            update_result = open(os.path.join(os.path.expanduser('~'), 'Marlin/.build_log')).read()
+            if 'No device matching following was found' in update_result:
+                self._logger.info("Failed update...")
+                self._update_status(False, "error", "A connected device was not found.")
+                self._clean_up()
+                break
+            elif 'FAILED' in update_result:
+                self._logger.info("Failed update...")
+                self._update_status(False, "error")
+                self._clean_up()
+                break
+            elif 'bytes of flash verified' in update_result and 'avrdude done' in update_result:
+                self._logger.info("Successful update!")
+                for line in update_result.splitlines():
+                    if "Reading" in line:
+                        self.read_time = self.find_between(line, " ", "s")
+                        self.completion_time += float(self.read_time)
+                    elif "Writing" in line:
+                        self.write_time = self.find_between(line, " ", "s")
+                        self.completion_time += float(self.write_time)
+
+                self._update_status(False, "completed", round(self.completion_time, 2))
+                self._clean_up()
+                break
+            elif 'ReceiveMessage(): timeout' in update_result:
+                self._logger.info("Update timed out. Check if port is already in use!")
+                self._update_status(False, "error", "Device timed out. Please check that the port is not in use!")
+
+                p = psutil.Process(self.updatePID)
+                for child in p.children(recursive=True):
+                    child.kill()
+                    p.kill()
+                self._clean_up()
+                break
+            sleep(1)
+
+    def _update_firmware_init(self, onstartup=False):
+        if self.printer_is_printing():
+            self._update_status(False, "error", "Printer is in use.")
+            raise RuntimeError("Printer is in use - cannot continue")
```

Comment:
Do we really want to raise an error here? Is there any cleanup we should be doing?